File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.py
class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        rows = len(grid)
        cols = len(grid[0])
        visited = [[False] * cols for _ in range(rows)]
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for row in range(rows):
            for col in range(cols):
                if visited[row][col]:
                    continue
                stack = [(row, col, -1, -1)]
                visited[row][col] = True
                while stack:
                    cr, cc, pr, pc = stack.pop()
                    for dr, dc in dirs:
                        nr, nc = cr + dr, cc + dc
                        if nr < 0 or nr >= rows or nc < 0 or nc >= cols:
                            continue
                        if grid[nr][nc] != grid[cr][cc]:
                            continue
                        if nr == pr and nc == pc:
                            continue
                        if visited[nr][nc]:
                            return True
                        visited[nr][nc] = True
                        stack.append((nr, nc, cr,cc))


        return False

File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.cpp
class Solution {
public:
    int find(int x, vector<int>& parent) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    bool union_sets(int a, int b, vector<int>& parent) {
        int ra = find(a, parent), rb = find(b, parent);
        if (ra == rb) 
            return true;
        parent[ra] = rb;
        return false;
    }

    bool containsCycle(vector<vector<char>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        vector<int> parent(rows * cols);
        for (int i = 0; i < rows * cols; i++) 
            parent[i] = i;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (j + 1 < cols && grid[i][j] == grid[i][j + 1]) {
                    if (union_sets(i * cols + j, i * cols + j + 1, parent)) 
                        return true;
                }
                if (i + 1 < rows && grid[i][j] == grid[i + 1][j]) {
                    if (union_sets(i * cols + j, (i + 1) * cols + j, parent)) 
                        return true;
                }
            }
        }
        return false;
    }
};